# КОММЕНТАРИЙ РЕВЬЮВЕРА
Привет Артем👋!)

Меня зовут Сибирев Владислав, я осуществлял проверку проекта. Может перейдем сразу на <b>ты</b>? Если это неприемлемо, то обязательно дай знать)

Далее по ходу работы ты сможешь встретить мои комментарии, они будут отмечены разными цветами:

<div class="alert alert-block alert-success">
 ✔️ Похвала и возможности развития
</div>
<div class="alert alert-block alert-warning">
⚠️ Некритичные замечания, но стоит обратить внимание
</div>
<div class="alert alert-block alert-danger">
❌ Грубая ошибка, необходимо исправить
</div>


Заранее прошу не удалять мои комментарии, это важно. Если у тебя возникнут вопросы, обозначь их, я с радостью тебе помогу!)

Как вариант, можешь обозначить их так:
<div class="alert alert-block alert-info">
<b>КОММЕНТАРИЙ СТУДЕНТА:</b>
</div>

# Исследование надежности заемщиков


Во второй части проекта вы выполните шаги 3 и 4. Их вручную проверит ревьюер.
Чтобы вам не пришлось писать код заново для шагов 1 и 2, мы добавили авторские решения в ячейки с кодом. 



## Откройте таблицу и изучите общую информацию о данных

**Задание 1. Импортируйте библиотеку pandas. Считайте данные из csv-файла в датафрейм и сохраните в переменную `data`. Путь к файлу:**

`/datasets/data.csv`

In [1]:
import pandas as pd

try:
    data = pd.read_csv('/datasets/data.csv')
except:
    data = pd.read_csv('https://code.s3.yandex.net/datasets/data.csv')

**Задание 2. Выведите первые 20 строчек датафрейма `data` на экран.**

In [2]:
data.head(20)

children  days_employed  dob_years            education  education_id  \
0          1   -8437.673028         42               высшее             0   
1          1   -4024.803754         36              среднее             1   
2          0   -5623.422610         33              Среднее             1   
3          3   -4124.747207         32              среднее             1   
4          0  340266.072047         53              среднее             1   
5          0    -926.185831         27               высшее             0   
6          0   -2879.202052         43               высшее             0   
7          0    -152.779569         50              СРЕДНЕЕ             1   
8          2   -6929.865299         35               ВЫСШЕЕ             0   
9          0   -2188.756445         41              среднее             1   
10         2   -4171.483647         36               высшее             0   
11         0    -792.701887         40              среднее             1   
12         0            NaN         65              среднее             1   
13         0   -1846.641941         54  неоконченное высшее             2   
14         0   -1844.956182         56               высшее             0   
15         1    -972.364419         26              среднее             1   
16         0   -1719.934226         35              среднее             1   
17         0   -2369.999720         33               высшее             0   
18         0  400281.136913         53              среднее             1   
19         0  -10038.818549         48              СРЕДНЕЕ             1   

       family_status  family_status_id gender income_type  debt  \
0    женат / замужем                 0      F   сотрудник     0   
1    женат / замужем                 0      F   сотрудник     0   
2    женат / замужем                 0      M   сотрудник     0   
3    женат / замужем                 0      M   сотрудник     0   
4   гражданский брак                 1      F   пенсионер     0   
5   гражданский брак                 1      M   компаньон     0   
6    женат / замужем                 0      F   компаньон     0   
7    женат / замужем                 0      M   сотрудник     0   
8   гражданский брак                 1      F   сотрудник     0   
9    женат / замужем                 0      M   сотрудник     0   
10   женат / замужем                 0      M   компаньон     0   
11   женат / замужем                 0      F   сотрудник     0   
12  гражданский брак                 1      M   пенсионер     0   
13   женат / замужем                 0      F   сотрудник     0   
14  гражданский брак                 1      F   компаньон     1   
15   женат / замужем                 0      F   сотрудник     0   
16   женат / замужем                 0      F   сотрудник     0   
17  гражданский брак                 1      M   сотрудник     0   
18    вдовец / вдова                 2      F   пенсионер     0   
19         в разводе                 3      F   сотрудник     0   

     total_income                                 purpose  
0   253875.639453                           покупка жилья  
1   112080.014102                 приобретение автомобиля  
2   145885.952297                           покупка жилья  
3   267628.550329              дополнительное образование  
4   158616.077870                         сыграть свадьбу  
5   255763.565419                           покупка жилья  
6   240525.971920                       операции с жильем  
7   135823.934197                             образование  
8    95856.832424                   на проведение свадьбы  
9   144425.938277                 покупка жилья для семьи  
10  113943.491460                    покупка недвижимости  
11   77069.234271       покупка коммерческой недвижимости  
12            NaN                         сыграть свадьбу  
13  130458.228857                 приобретение автомобиля  
14  165127.911772              покупка жилой недвижимости  
15  116820.904450 

**Задание 3. Выведите основную информацию о датафрейме с помощью метода `info()`.**

In [3]:
data.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 21525 entries, 0 to 21524
Data columns (total 12 columns):
 #   Column            Non-Null Count  Dtype  
---  ------            --------------  -----  
 0   children          21525 non-null  int64  
 1   days_employed     19351 non-null  float64
 2   dob_years         21525 non-null  int64  
 3   education         21525 non-null  object 
 4   education_id      21525 non-null  int64  
 5   family_status     21525 non-null  object 
 6   family_status_id  21525 non-null  int64  
 7   gender            21525 non-null  object 
 8   income_type       21525 non-null  object 
 9   debt              21525 non-null  int64  
 10  total_income      19351 non-null  float64
 11  purpose           21525 non-null  object 
dtypes: float64(2), int64(5), object(5)
memory usage: 2.0+ MB


## Предобработка данных

### Удаление пропусков

**Задание 4. Выведите количество пропущенных значений для каждого столбца. Используйте комбинацию двух методов.**

In [4]:
data.isna().sum()

children               0
days_employed       2174
dob_years              0
education              0
education_id           0
family_status          0
family_status_id       0
gender                 0
income_type            0
debt                   0
total_income        2174
purpose                0
dtype: int64

**Задание 5. В двух столбцах есть пропущенные значения. Один из них — `days_employed`. Пропуски в этом столбце вы обработаете на следующем этапе. Другой столбец с пропущенными значениями — `total_income` — хранит данные о доходах. На сумму дохода сильнее всего влияет тип занятости, поэтому заполнить пропуски в этом столбце нужно медианным значением по каждому типу из столбца `income_type`. Например, у человека с типом занятости `сотрудник` пропуск в столбце `total_income` должен быть заполнен медианным доходом среди всех записей с тем же типом.**

In [5]:
for t in data['income_type'].unique():
    data.loc[(data['income_type'] == t) & (data['total_income'].isna()), 'total_income'] = \
    data.loc[(data['income_type'] == t), 'total_income'].median()

### Обработка аномальных значений

**Задание 6. В данных могут встречаться артефакты (аномалии) — значения, которые не отражают действительность и появились по какой-то ошибке. таким артефактом будет отрицательное количество дней трудового стажа в столбце `days_employed`. Для реальных данных это нормально. Обработайте значения в этом столбце: замените все отрицательные значения положительными с помощью метода `abs()`.**

In [6]:
data['days_employed'] = data['days_employed'].abs()

**Задание 7. Для каждого типа занятости выведите медианное значение трудового стажа `days_employed` в днях.**

In [7]:
data.groupby('income_type')['days_employed'].agg('median')

income_type
безработный        366413.652744
в декрете            3296.759962
госслужащий          2689.368353
компаньон            1547.382223
пенсионер          365213.306266
предприниматель       520.848083
сотрудник            1574.202821
студент               578.751554
Name: days_employed, dtype: float64

У двух типов (безработные и пенсионеры) получатся аномально большие значения. Исправить такие значения сложно, поэтому оставьте их как есть. Тем более этот столбец не понадобится вам для исследования.

**Задание 8. Выведите перечень уникальных значений столбца `children`.**

In [8]:
data['children'].unique()

array([ 1,  0,  3,  2, -1,  4, 20,  5])

**Задание 9. В столбце `children` есть два аномальных значения. Удалите строки, в которых встречаются такие аномальные значения из датафрейма `data`.**

In [9]:
data = data[(data['children'] != -1) & (data['children'] != 20)]

**Задание 10. Ещё раз выведите перечень уникальных значений столбца `children`, чтобы убедиться, что артефакты удалены.**

In [10]:
data['children'].unique()

array([1, 0, 3, 2, 4, 5])

### Удаление пропусков (продолжение)

**Задание 11. Заполните пропуски в столбце `days_employed` медианными значениями по каждого типа занятости `income_type`.**

In [11]:
for t in data['income_type'].unique():
    data.loc[(data['income_type'] == t) & (data['days_employed'].isna()), 'days_employed'] = \
    data.loc[(data['income_type'] == t), 'days_employed'].median()

**Задание 12. Убедитесь, что все пропуски заполнены. Проверьте себя и ещё раз выведите количество пропущенных значений для каждого столбца с помощью двух методов.**

In [12]:
data.isna().sum()

children            0
days_employed       0
dob_years           0
education           0
education_id        0
family_status       0
family_status_id    0
gender              0
income_type         0
debt                0
total_income        0
purpose             0
dtype: int64

### Изменение типов данных

**Задание 13. Замените вещественный тип данных в столбце `total_income` на целочисленный с помощью метода `astype()`.**

In [13]:
data['total_income'] = data['total_income'].astype(int)

### Обработка дубликатов

**Задание 14. Обработайте неявные дубликаты в столбце `education`. В этом столбце есть одни и те же значения, но записанные по-разному: с использованием заглавных и строчных букв. Приведите их к нижнему регистру. Проверьте остальные столбцы.**

In [14]:
data['education'] = data['education'].str.lower()

**Задание 15. Выведите на экран количество строк-дубликатов в данных. Если такие строки присутствуют, удалите их.**

In [15]:
data.duplicated().sum()

71

In [16]:
data = data.drop_duplicates()

### Категоризация данных

**Задание 16. На основании диапазонов, указанных ниже, создайте в датафрейме `data` столбец `total_income_category` с категориями:**

- 0–30000 — `'E'`;
- 30001–50000 — `'D'`;
- 50001–200000 — `'C'`;
- 200001–1000000 — `'B'`;
- 1000001 и выше — `'A'`.


**Например, кредитополучателю с доходом 25000 нужно назначить категорию `'E'`, а клиенту, получающему 235000, — `'B'`. Используйте собственную функцию с именем `categorize_income()` и метод `apply()`.**

In [17]:
def categorize_income(income):
    try:
        if 0 <= income <= 30000:
            return 'E'
        elif 30001 <= income <= 50000:
            return 'D'
        elif 50001 <= income <= 200000:
            return 'C'
        elif 200001 <= income <= 1000000:
            return 'B'
        elif income >= 1000001:
            return 'A'
    except:
        pass

In [18]:
data['total_income_category'] = data['total_income'].apply(categorize_income)

**Задание 17. Выведите на экран перечень уникальных целей взятия кредита из столбца `purpose`.**

In [19]:
data['purpose'].unique()

array(['покупка жилья', 'приобретение автомобиля',
       'дополнительное образование', 'сыграть свадьбу',
       'операции с жильем', 'образование', 'на проведение свадьбы',
       'покупка жилья для семьи', 'покупка недвижимости',
       'покупка коммерческой недвижимости', 'покупка жилой недвижимости',
       'строительство собственной недвижимости', 'недвижимость',
       'строительство недвижимости', 'на покупку подержанного автомобиля',
       'на покупку своего автомобиля',
       'операции с коммерческой недвижимостью',
       'строительство жилой недвижимости', 'жилье',
       'операции со своей недвижимостью', 'автомобили',
       'заняться образованием', 'сделка с подержанным автомобилем',
       'получение образования', 'автомобиль', 'свадьба',
       'получение дополнительного образования', 'покупка своего жилья',
       'операции с недвижимостью', 'получение высшего образования',
       'свой автомобиль', 'сделка с автомобилем',
       'профильное образование', 'высшее об

**Задание 18. Создайте функцию, которая на основании данных из столбца `purpose` сформирует новый столбец `purpose_category`, в который войдут следующие категории:**

- `'операции с автомобилем'`,
- `'операции с недвижимостью'`,
- `'проведение свадьбы'`,
- `'получение образования'`.

**Например, если в столбце `purpose` находится подстрока `'на покупку автомобиля'`, то в столбце `purpose_category` должна появиться строка `'операции с автомобилем'`.**

**Используйте собственную функцию с именем `categorize_purpose()` и метод `apply()`. Изучите данные в столбце `purpose` и определите, какие подстроки помогут вам правильно определить категорию.**

In [20]:
def categorize_purpose(row):
    try:
        if 'автом' in row:
            return 'операции с автомобилем'
        elif 'жил' in row or 'недвиж' in row:
            return 'операции с недвижимостью'
        elif 'свад' in row:
            return 'проведение свадьбы'
        elif 'образов' in row:
            return 'получение образования'
    except:
        return 'нет категории'

In [21]:
data['purpose_category'] = data['purpose'].apply(categorize_purpose)

### Шаг 3. Исследуйте данные и ответьте на вопросы

#### 3.1 Есть ли зависимость между количеством детей и возвратом кредита в срок?

In [31]:
# Ваш код будет здесь. Вы можете создавать новые ячейки.
#mydata1 = data.loc[data['debt'] == 1]['children']
#mydata1 = mydata1.value_counts(sort = False).reset_index(name = 'debtors')
#mydata1 = mydata1.rename(columns = {'index':'children'})
#mydata2 = data.loc[data['debt'] == 0]['children']
#mydata2 = mydata2.value_counts(sort = False).reset_index(name = 'non-debtors')
#mydata2 = mydata2.rename(columns = {'index':'children'})
#ratio = pd.DataFrame(columns={'children', 'ratio'})
#ratio['children'] = mydata1['children']
#ratio['ratio'] = mydata1['debtors'] / mydata2['non-debtors']
#print('Amount of debtors with corresponding amount of children\n', mydata1, '\n')
#print('Amount of people without a debt with corresponding amount of children\n',mydata2, '\n')
#print('Ratio of the number of debtors to the number of people without a debt')
#print(ratio)
table = pd.pivot_table(data, index=['children'], values=['debt'], aggfunc={'sum','count','mean'})
table.loc[:,'debt']['mean']*=100
table = table.reset_index(drop=False,col_fill='children').droplevel(0, axis=1)
table = table.rename(columns={'count':'amount of people in this category', 'mean':'percentage of debtors', 'sum':'debtors'})
table
    
    

children  amount of people in this category  percentage of debtors  debtors
0         0                            14091.0               7.543822   1063.0
1         1                             4808.0               9.234609    444.0
2         2                             2052.0               9.454191    194.0
3         3                              330.0               8.181818     27.0
4         4                               41.0               9.756098      4.0
5         5                                9.0               0.000000      0.0

<div class="alert alert-block alert-warning">
<b>Комментарий ревьюер:</b> 

⚠️ Здесь, молодец, логика правильная, но давай все данные объединим в одну таблицу. Что здесь нужно сделать (это касается изучение всех взаимосвязей):
1) Сделать одну таблицу, используй для этого `pivot_table`, индекс это будут дети, значение должники, а функции будут `sum (количество должников), count (количество кредиторов), mean (доля должников)`.<br>
2) Столбец `mean` можно умножить на 100<br>
3) Для красоты лучше избавляться от мультииндексов, после сброса `reset_index()` поставь `droplevel(1, axis=1)`, названия столбцов сразу же станут красивыми;<br>
P.S. С помощью `Print()`  не нужно выводить таблицу, просто название прописываешь и все выведется<br>

<div class="alert alert-block alert-success">
<b>Комментарий ревьюера_V2:</b>

✔️  Здесь все отлично стало)) Все замечания исправил)) 

  </div>

**Вывод:** 
Люди без детей лучше всего отдают кредит, доля должников здесь составляет 7,5%. Несколько хуже все у группы с 1-м и 2-мя детьми: доля невозврата составляет соответственно 9,2% и 9,4%. Остальные категории уже некорректно сравнивать, так как в них очень маленькая выборка.

<div class="alert alert-block alert-warning">
<b>Комментарий ревьюер:</b> 

⚠️ Не бойся использовать в выводах цифры, также нужно сделать акцент на сбалансированности выборок. Примерно выводы будут выглядеть так: 
`люди, которые не имеют детей, лучше всего отдают кредит, доля должников здесь составляет с 7,5%, хуже обстоит дело с должниками, кто имеет одного или двух детей, здесь доля невозврата составляет соответственно 9,2% и 9,4%. Остальные категории уже некорректно сравнивать, так как в них очень маленькая выборка`)) <br>

<div class="alert alert-block alert-success">
<b>Комментарий ревьюера_V2:</b>

✔️  Выводы переработал, молодец)) 

  </div>

#### 3.2 Есть ли зависимость между семейным положением и возвратом кредита в срок?

In [32]:
# Ваш код будет здесь. Вы можете создавать новые ячейки.
#mydata1 = data.loc[data['debt'] == 1]['family_status']
#mydata1 = mydata1.value_counts().reset_index(name = 'debtors')
#mydata1 = mydata1.rename(columns = {'index':'family_status'})
#mydata2 = data.loc[data['debt'] == 0]['family_status']
#mydata2 = mydata2.value_counts().reset_index(name = 'non-debtors')
#mydata2 = mydata2.rename(columns = {'index':'family_status'})
#ratio = pd.DataFrame(columns={'family_status', 'ratio'})
#ratio['family_status'] = mydata1['family_status']
#ratio['ratio'] = mydata1['debtors'] / mydata2['non-debtors']
#print('Amount of debtors with corresponding family status\n', mydata1, '\n')
#print('Amount of people without a debt with corresponding family status\n',mydata2, '\n')
#print('Ratio of the number of debtors to the number of people without a debt')
#print(ratio)
table = pd.pivot_table(data, index=['family_status'], values=['debt'], aggfunc={'sum','count','mean'})
table.loc[:,'debt']['mean']*=100
table = table.reset_index(drop=False,col_fill='family_status').droplevel(0, axis=1)
table = table.rename(columns={'count':'amount of people in this category', 'mean':'percentage of debtors', 'sum':'debtors'})
table
    


family_status  amount of people in this category  \
0  Не женат / не замужем                             2796.0   
1              в разводе                             1189.0   
2         вдовец / вдова                              951.0   
3       гражданский брак                             4134.0   
4        женат / замужем                            12261.0   

   percentage of debtors  debtors  
0               9.763948    273.0  
1               7.064760     84.0  
2               6.624606     63.0  
3               9.313014    385.0  
4               7.560558    927.0

**Вывод:** Доля должников среди женатых/замужних людей составляет 7,56%. Хуже все с людьми в гражданском браке - там этот показатель составляет целых 9,31%. Еще хуже все у не женатых/не замужних - 9,76%. У остальных категорий выборка слишком мала относительно категории "женат/замужем", так что сравнивать доли будет некорректно.

<div class="alert alert-block alert-warning">
<b>Комментарий ревьюер:</b> 

⚠️ В выводах давай будет использовать цифры, также  удели здесь внимание тоже сбалансированности выборок))

<div class="alert alert-block alert-success">
<b>Комментарий ревьюера_V2:</b>

✔️  Здесь тоже все отлично по выводам))

  </div>

#### 3.3 Есть ли зависимость между уровнем дохода и возвратом кредита в срок?

In [35]:
# Ваш код будет здесь. Вы можете создавать новые ячейки.
#mydata1 = data.loc[data['debt'] == 1]['total_income_category']
#mydata1 = mydata1.value_counts().reset_index(name='debtors').sort_values(by='index')
#mydata1 = mydata1.rename(columns = {'index':'total_income_category'})
#mydata1.reset_index(drop=True,inplace=True)
#mydata2 = data.loc[data['debt'] == 0]['total_income_category']
#mydata2 = mydata2.value_counts().reset_index(name='non-debtors').sort_values(by='index')
#mydata2 = mydata2.rename(columns = {'index':'total_income_category'})
#mydata2.reset_index(drop=True,inplace=True)
#ratio = pd.DataFrame(columns={'total_income_category', 'ratio'})
#ratio['total_income_category'] = mydata1['total_income_category']
#ratio['ratio'] = mydata1['debtors'] / mydata2['non-debtors']
#print('Amount of debtors with corresponding income category\n', mydata1, '\n')
#print('Amount of people without a debt with corresponding income category\n',mydata2, '\n')
#print('Ratio of the number of debtors to the number of people without a debt')
#print(ratio)
table = pd.pivot_table(data, index=['total_income_category'], values=['debt'], aggfunc={'sum','count','mean'})
table.loc[:,'debt']['mean']*=100
table = table.reset_index(drop=False,col_fill='total_income_category').droplevel(0, axis=1)
table = table.rename(columns={'count':'amount of people in this category', 'mean':'percentage of debtors', 'sum':'debtors'})
table
    

total_income_category  amount of people in this category  \
0                     A                               25.0   
1                     B                             5014.0   
2                     C                            15921.0   
3                     D                              349.0   
4                     E                               22.0   

   percentage of debtors  debtors  
0               8.000000      2.0  
1               7.060231    354.0  
2               8.498210   1353.0  
3               6.017192     21.0  
4               9.090909      2.0

**Вывод:** Так как выборка должна быть сбалансированной, то корректно будет сравнивать только категроии B(200001-1000000) и C(50001-200000). В категории B доля должников составляет 7,06%, в то время как в категории C этот показатель существенно выше и составялет 8,5%. Из чего можно сделать вывод, что люди из категории B более надежные.

<div class="alert alert-block alert-warning">
<b>Комментарий ревьюер:</b> 

⚠️ Здесь сравниваем только две группы B и С, не забываем про сбалансированость выборки))<br>

<div class="alert alert-block alert-success">
<b>Комментарий ревьюера_V2:</b>

✔️  Здесь сравнил две самые большие группы))

  </div>

#### 3.4 Как разные цели кредита влияют на его возврат в срок?

In [34]:
# Ваш код будет здесь. Вы можете создавать новые ячейки.
#mydata1 = data.loc[data['debt'] == 1]['purpose_category']
#mydata1 = mydata1.value_counts().reset_index(name='debtors').sort_values(by='index')
#mydata1 = mydata1.rename(columns = {'index':'purpose_category'})
#mydata1.reset_index(drop=True,inplace=True)
#mydata2 = data.loc[data['debt'] == 0]['purpose_category']
#mydata2 = mydata2.value_counts().reset_index(name='non-debtors').sort_values(by='index')
#mydata2 = mydata2.rename(columns = {'index':'purpose_category'})
#mydata2.reset_index(drop=True,inplace=True)
#ratio = pd.DataFrame(columns={'purpose_category', 'ratio'})
#ratio['purpose_category'] = mydata1['purpose_category']
#ratio['ratio'] = mydata1['debtors'] / mydata2['non-debtors']
#print('Amount of debtors with corresponding purpose category\n', mydata1, '\n')
#print('Amount of people without a debt with corresponding purpose category\n',mydata2, '\n')
#print('Ratio of the number of debtors to the number of people without a debt')
#print(ratio)
table = pd.pivot_table(data, index=['purpose_category'], values=['debt'], aggfunc={'sum','count','mean'})
table.loc[:,'debt']['mean']*=100
table = table.reset_index(drop=False,col_fill='purpose_category').droplevel(0, axis=1)
table = table.rename(columns={'count':'amount of people in this category', 'mean':'percentage of debtors', 'sum':'debtors'})
table
    

purpose_category  amount of people in this category  \
0    операции с автомобилем                             4279.0   
1  операции с недвижимостью                            10751.0   
2     получение образования                             3988.0   
3        проведение свадьбы                             2313.0   

   percentage of debtors  debtors  
0               9.347978    400.0  
1               7.255139    780.0  
2               9.252758    369.0  
3               7.911803    183.0

**Вывод:** Я не вижу проблемы: процент должников наименьший(7,25%) в категории "операции с недвижимостью" => люди, берущие кредить именно на эти операции являются самыми надежными. Несколько хуже c операциями на проведение свадеб: 7,9%. Примерно одинаково плохо все в категориях "операции с автомобилем" и "получение образования" - там процент должников составялет 9,34% и 9,25% соответственно.

<div class="alert alert-block alert-warning">
<b>Комментарий ревьюера:</b> 

⚠️ Почему по твоему самые надежные заемщики те, кто берет ипотеку? Здесь можно порасуждать в выводах)) 

<div class="alert alert-block alert-success">
<b>Комментарий ревьюера_V2:</b>

✔️ Самые большие суммы при ипотеке, поэтому банки предъявляют самые жесткие требования к ипотечникам, поэтому их проверяют вдоль и поперек при выдаче кредитов, плюс собственность находится в залоге у банка, не будешь платить она вернется к банку)) 
   
       
  </div>

#### 3.5 Приведите возможные причины появления пропусков в исходных данных.

*Ответ:* Пропуски содержались в двух столбцах - "трудовой стаж" и "ежемесячный доход". Скорее всего пропуски появились не в результате ошибки. Такие люди либо пенсионеры, либо безработные. Так как количество пропусков в обоих столбцах совпадает, то вероятность ошибки очень мала.

<div class="alert alert-block alert-warning">
<b>Комментарий ревьюера:</b>

⚠️ Все верно, если пристально взглянуть на на наши данные, то увидим что пропуски в столбцах "трудовой стаж" и "ежемесячный доход" абсолютно идентичны. Количество ячеек с отсутствующими данными в обоих указанных столбцах совпадает и равно 2174. Почему такое могло возникнуть, еси мы не будем учитывать только человеческий фактор?)

<div class="alert alert-block alert-success">
<b>Комментарий ревьюера_V2:</b>

✔️ Да, всё здраво. В принципе эти возможные причины сразу и приходят в голову. 
    
Однако, если пристально взглянуть на на наши данные, то увидим что пропуски в столбцах "трудовой стаж" и "ежемесячный доход" абсолютно идентичны. Количество ячеек с отсутствующими данными в обоих указанных столбцах совпадает и равно 2174. Маловероятно, что специалисты по выдаче кредитов, заполнив все без исключения ячейки других столбцов по каждому кредитополучателю, оставляли незаполненными именно два показателя, и сделали это ровно 2174 раза в каждом столбце. Даже не проводя проверку, во всех ли случаях незаполненные ячейки находятся в одной и то же строке с высокой долей вероятности можно утверждать, что произошла техническая ошибка при формировании таблицы с исходными данными (либо некий сбой во время выгрузки данных на анализ, либо при объединении данных в единый файл, который также должен был передаваться на анализ, либо эти данные не собирались).
</div>

#### 3.6 Объясните, почему заполнить пропуски медианным значением — лучшее решение для количественных переменных.

*Ответ:* Потому что в таком случае средняя зарплата и количество проработанных дней изменятся минимально, а мы не потеряем данные о пенсионерах и безработных. Медиана - это число, которое находится в середине некоторого набора. Если числа например 1,2,3,4,100 - то медианой будет число 3, а вот средним будет 22 - разница существенная.

<div class="alert alert-block alert-warning">
<b>Комментарий ревьюера:</b> 

⚠️ А в чем разница между средним и медианой?))

<div class="alert alert-block alert-success">
<b>Комментарий ревьюера_V2:</b>

✔️ Медиана не подвержена выбросам, все верно)) 

### Шаг 4: общий вывод.

1. Зависимость между количество детей и возвратом кредита в срок есть: процент должников увеличивается с ростом числа детей(0 - 7,5%, 1 - 9,2%, 2 - 9,4%. Остальные выборки слишком малы.
2. Зависимость между семейным положением и возвратом кредита в срок также есть: чем серьезнее у людей отношения с партнером, тем лучше они возвращают долги: доля должников у не женатых/не замужних 9,8%, в группе "гражданский брак" 9,3%, а вот у женатых/замужних 7,6%. Остальные выборки слишком малы.
3. Зависимость между уровнем дохода и возвратом кредита в срок тоже имеется: люди с бОльшим доходом лучше возвращают займ. У категории B(200001-1000000) 7,1%, а у категории C(50001-200000) 8,5%.
4. Процент должников наименьший(7,25%) в категории "операции с недвижимостью" => люди, берущие кредить именно на эти операции являются самыми надежными. Несколько хуже c операциями на проведение свадеб: 7,9%. Примерно одинаково плохо все в категориях "операции с автомобилем" и "получение образования" - там процент должников составялет 9,34% и 9,25% соответственно.
- Необходимо собирать больше данных для категорий, где выборка сравнительно невелика(например, люди с 3-мя, 4-мя, 5-ю детьми).


 <div class="alert alert-block alert-success">
<b>Комментарий ревьюера_V2:</b>

✔️ Выводы хорошо переработал, молодец)) Рекомендации даны банку, исследование проведено))
  
  
</div>

<div class="alert alert-block alert-danger">
<b>Комментарий ревьюера:</b> 

❌ Выводы должны быть детальными:
- Отвечаешь на вопрос "Есть ли зависимость между количеством детей и возвратом кредита в срок?"
- Отвечаешь на вопрос  "Есть ли зависимость между семейным положением и возвратом кредита в срок?"
- Отвечаешь на вопрос  "Есть ли зависимость между уровнем дохода и возвратом кредита в срок?"
- Отвечаешь на вопрос  "Как разные цели кредита влияют на его возврат в срок?"
- Даешь рекомендацией банку по улучшению сбора данных и скоринга.

Общий вывод - резюме этой работы. Большинство заказчиков посмотрят сначала его (а некоторые только его). Чтобы выводы не огорошивали заказчика своей очевидностью, стоит добавить в них цифр, всяких корреляций, процентов итд :) Сделай так, чтобы вся найденная в данных информация была отражена в выводе))

<div class="alert alert-block alert-success">
<b>Комментарий ревьюера:</b>
    
Артем, отлично поработал, видно что умеешь работать с кодом,  в общем молодец, но нужно будет оптимизировать код при изучении взаимосвязей и поработать над выводами)) 
<br> Жду исправленного варианта)
  </div>

<div class="alert alert-block alert-success">
<b>Комментарий ревьюера_V2:</b>
    
Артем,  отлично поработал, можно двигаться дальше)) Успехов в дальнейшем обучении)) 
  </div>